htlcswitch/switch.go

```diff
@@ -205,12 +190,18 @@ type Switch struct {
 	// service was initialized with.
 	cfg *Config
 
-	// pendingPayments stores payments initiated by the user that are not yet
-	// settled. The map is used to later look up the payments and notify the
-	// user of the result when they are complete. Each payment is given a unique
-	// integer ID when it is created.
-	pendingPayments map[uint64]*pendingPayment
-	pendingMutex    sync.RWMutex
+	// paymentResults stores payments initiated by the user, and their
+	// results.  The store is used to later look up the payments and notify
+	// the user of the result when they are complete. Each payment attempt
+	// should be given a unique integer ID when it is created, and only
+	// added once.
+	paymentResults *paymentResultStore
+
+	// deobfuscators is a map storing error decrypters we need to parse
+	// received payment failures. We don't store these in the database, as
+	// being unable to decrypt an error after restart is not critical.
+	deobfuscators    map[uint64]ErrorDecrypter
```

Comment:
How consistent is it that we encode the onion in the router and decode it in the switch? I tend to think it may be better to also move the decoding to the router.

The router already has the route stored, so to decode the failure it would only need to store the session key as an extra field in the control tower db.

This will also fix the problem that we can't decode a failure after restart of the switch.

Reply:
Fixed by #3063